### PR TITLE
RadosClient: result code may overflow

### PR DESCRIPTION
**Pull request by xiexingguo**
For more details about this issue, check http://tracker.ceph.com/issues/13715

### CODE REVIEW

**Comment by xiexingguo** [comment]
@tchaikov  Any suggestions?

**Comment by tchaikov** [comment]
@xiexingguo i am not sure this is the right fix. probably we should deprecate these two methods and have new methods overloading them, for example:
```C++
int pool_requires_alignment(int64_t pool_id, bool *requires);
```
by returning the error code, and pass a pointer for the "[out]" parameter?

@liewegas and @dillaman what do you think?

**Comment by tchaikov** [comment]
and your `Signed-off-by` does not look right: missing angle brackets.

**Comment by xiexingguo** [comment]
@tchaikov 
It is one of the possible resolutions if you please refer to http://tracker.ceph.com/issues/13715.
I'll get ready to work on it once i get confirmation from you guys.

**Comment by liewegas** [comment]
Agree on fixing the function prototypes to have a return code separate from the output arguments.

**Comment by dillaman** [comment]
@tchaikov I like the new function idea since it's safer.  However, to preserve API compatibility, the new function would need to be named:

```
int pool_requires_alignment2(int64_t pool_id, bool *requires);
```

**Comment by dachary** [comment]
@xiexingguo please ignore the false negative from the bot

**Comment by xiexingguo** [comment]
@dillaman Would you please review this change for me? Thanks a lot.

**Comment by xiexingguo** [comment]
@tchaikov  Fixup, please review.

File: src/librados/RadosClient.cc [hunk 1]
```diff
@@ -106,6 +106,28 @@ bool librados::RadosClient::pool_requires_alignment(int64_t pool_id)
...
+    return -ENOENT;
+  }
+  *requires = osdmap->get_pg_pool(pool_id)->requires_aligned_append();
+  objecter->put_osdmap_read();
+  return 0;
+}
+
 uint64_t librados::RadosClient::pool_required_alignment(int64_t pool_id)
 {
   int r = wait_for_osdmap();
```
Comment (tchaikov) [inline src/librados/RadosClient.cc]: @xiexingguo could you reimplement the deprecated methods using the new ones in another commit in this pr? so we don't less duplicated code and the logic reveals the problem of the old methods in a more visible way.

Comment (xiexingguo) [inline src/librados/RadosClient.cc]: Sorry, I don't quite understand what do you mean. Can you make an example?

Comment (tchaikov) [inline src/librados/RadosClient.cc]: @xiexingguo 
```patch
diff --git a/src/librados/RadosClient.cc b/src/librados/RadosClient.cc
index a01ae7d..6157e35 100644
--- a/src/librados/RadosClient.cc
+++ b/src/librados/RadosClient.cc
@@ -108,16 +108,13 @@ bool librados::RadosClient::pool_requires_alignment(int64_t pool_id)
 
 uint64_t librados::RadosClient::pool_required_alignment(int64_t pool_id)
 {
-  int r = wait_for_osdmap();
+  uint64_t alignment = 0;
+  int r = pool_required_alignment2(pool_id, &alignment);
   if (r < 0) {
     return r;
+  } else {
+    return alignment;
   }
-
-  const OSDMap *osdmap = objecter->get_osdmap_read();
-  uint64_t ret = osdmap->have_pg_pool(pool_id) ?
-    osdmap->get_pg_pool(pool_id)->required_alignment() : 0;
-  objecter->put_osdmap_read();
-  return ret;
 }
 
```

Comment (xiexingguo) [inline src/librados/RadosClient.cc]: @tchaikov  Got it. I will do it now.

**Comment by xiexingguo** [comment]
@tchaikov UT passed, please review.

**Comment by tchaikov** [comment]
lgtm.

**Comment by xiexingguo** [comment]
@liewegas Hi, sage, could you please add this one for test too?

File: src/librados/RadosClient.cc [hunk 2]
```diff
@@ -94,30 +94,68 @@ int64_t librados::RadosClient::lookup_pool(const char *name)
...
-  return ret;
+  return 0;
 }
 
 uint64_t librados::RadosClient::pool_required_alignment(int64_t pool_id)
 {
+  uint64_t alignment;
+  int r = pool_required_alignment2(pool_id, &alignment);
+  if (r < 0) {
+    return r;
```
Comment (dzafman) [inline src/librados/RadosClient.cc]: One problem here.  Return a negative int as uint64_t.  What does caller do with a huge number?  Should it return 0 and caller should check for that as an error indication?

Comment (dzafman) [inline src/librados/RadosClient.cc]: I see that is why there is a new interface, but I think returning 0 on error for the original interface is reasonable.

**Comment by xiexingguo** [comment]
@dzafman Fixup as you suggested. Please review.

**Comment by dzafman** [comment]
lgtm

**Pull request merged**